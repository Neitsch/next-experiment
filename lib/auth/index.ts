import Cookie from "js-cookie";
import jwtDecode from "jwt-decode";

const getQueryParams = () => {
  const params = {};
  window.location.href.replace(
    /([^(?|#)=&]+)(=([^&]*))?/g,
    ($0, $1, $2, $3) => {
      params[$1] = $3;
    },
  );
  return params;
};

export const extractInfoFromHash = () => {
  if (!process.browser) {
    return undefined;
  }
  const { id_token, access_token, state, expires_in } = getQueryParams();
  return { token: id_token, access_token, secret: state, expires_in };
};

export const setToken = (token, accessToken, expiresIn) => {
  if (!process.browser) {
    return;
  }
  const expires = new Date();
  expires.setSeconds(expires.getSeconds() + expiresIn);
  Cookie.set("user", jwtDecode(token), { expires, SameSite: "Lax" });
  Cookie.set("jwt", token, { expires, SameSite: "Lax" });
  Cookie.set("token", accessToken, { expires, SameSite: "Lax" });
};

export const unsetToken = () => {
  if (!process.browser) {
    return;
  }
  Cookie.remove("jwt");
  Cookie.remove("user");
  Cookie.remove("secret");

  // to support logging out from all windows
  window.localStorage.setItem("logout", Date.now());
};

export const getUserFromServerCookie = req => {
  const cookie = getServerCookie(req);
  if (!cookie) {
    return undefined;
  }
  return jwtDecode(cookie);
};

export const getServerCookie = req => {
  if (!req.headers.cookie) {
    return undefined;
  }
  const jwtCookie = req.headers.cookie
    .split(";")
    .find(c => c.trim().startsWith("jwt="));
  if (!jwtCookie) {
    return undefined;
  }
  return jwtCookie.split("=")[1];
};

export const getLocalCookie = () => {
  return Cookie.getJSON("jwt");
};

export const getUserFromLocalCookie = () => {
  return Cookie.getJSON("user");
};

export const setSecret = secret =>
  Cookie.set("secret", secret, { SameSite: "Lax" });

export const checkSecret = secret => {
  return Cookie.get("secret") === secret;
};
