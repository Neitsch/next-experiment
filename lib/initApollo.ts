import { Hermes } from "apollo-cache-hermes";
import { persistCache } from "apollo-cache-persist";
import { ApolloClient } from "apollo-client";
import { setContext } from "apollo-link-context";
import { HttpLink } from "apollo-link-http";
import { createPersistedQueryLink } from "apollo-link-persisted-queries";
import isomorphicUnfetch from "isomorphic-unfetch";
import localForage from "localforage";

let apolloClient = null;

export const setAuthorizationLink = jwt =>
  setContext(() => ({
    headers: {
      authorization: "Bearer " + jwt,
    },
  }));

export const setupLink = jwt =>
  setAuthorizationLink(jwt).concat(
    createPersistedQueryLink().concat(
      new HttpLink({
        credentials: "same-origin",
        fetch: isomorphicUnfetch,
        uri: "http://localhost:3000/graphql",
      }),
    ),
  );

function create(jwt) {
  const cache = new Hermes({});
  if (process.browser) {
    persistCache({
      cache,
      storage: localForage,
    });
  }
  return new ApolloClient({
    cache,
    connectToDevTools: !!process.browser,
    link: setAuthorizationLink(jwt).concat(
      createPersistedQueryLink().concat(
        new HttpLink({
          credentials: "same-origin",
          fetch: isomorphicUnfetch,
          uri: "http://localhost:3000/graphql",
        }),
      ),
    ),
    ssrMode: !process.browser,
  });
}

export const resetClient = () => {
  apolloClient = null;
};

export default function initApollo(jwt) {
  if (!process.browser) {
    return create(jwt);
  }
  if (!apolloClient) {
    apolloClient = create(jwt);
  }
  return apolloClient;
}
