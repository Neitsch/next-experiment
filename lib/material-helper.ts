/* tslint:disable-next-line */
import green from "@material-ui/core/colors/green";
/* tslint:disable-next-line */
import purple from "@material-ui/core/colors/purple";
import {
  createGenerateClassName,
  createMuiTheme,
  /* tslint:disable-next-line */
} from "@material-ui/core/styles";
import { SheetsRegistry } from "jss";

const theme = createMuiTheme({
  palette: {
    primary: {
      dark: purple[700],
      light: purple[300],
      main: purple[500],
    },
    secondary: {
      dark: green[700],
      light: green[300],
      main: green[500],
    },
  },
});

function createPageContext() {
  return {
    generateClassName: createGenerateClassName(),
    sheetsManager: new Map(),
    sheetsRegistry: new SheetsRegistry(),
    theme,
  };
}

export function getPageContext() {
  if (!process.browser) {
    return createPageContext();
  }

  // Reuse context on the client-side.
  if (!global.__INIT_MATERIAL_UI__) {
    global.__INIT_MATERIAL_UI__ = createPageContext();
  }

  return global.__INIT_MATERIAL_UI__;
}
