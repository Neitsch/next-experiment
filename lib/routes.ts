import Routes from "next-routes";

const routes = new Routes().add("index", "/");
export default routes;
export const Link = routes.Link;
export const Router = routes.Router;
