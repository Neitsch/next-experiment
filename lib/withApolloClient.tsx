import { ApolloClient } from "apollo-client";
/* tslint:disable-next-line */
import Head from "next/head";
import React from "react";
import { getDataFromTree } from "react-apollo";

import initApollo from "./initApollo";

export default App => {
  return class Apollo extends React.Component {
    public static displayName = "withApollo(App)";
    public static async getInitialProps(ctx) {
      const { Component, router } = ctx;
      let appProps = {};
      if (App.getInitialProps) {
        appProps = await App.getInitialProps(ctx);
      }
      const apolloState = {
        data: {},
      };
      const apollo = initApollo({});
      try {
        await getDataFromTree(
          <App
            {...appProps}
            Component={Component}
            router={router}
            apolloState={apolloState}
            apolloClient={apollo}
          />,
        );
      } catch (error) {
        /* tslint:disable-next-line */
        console.error("Error while running `getDataFromTree`", error);
      }

      if (!process.browser) {
        Head.rewind();
      }
      apolloState.data = apollo.cache.extract();

      return {
        ...appProps,
        apolloState,
      };
    }

    private apolloClient: ApolloClient<any>;

    constructor(props) {
      super(props);
      this.apolloClient =
        props.apolloClient || initApollo(props.apolloState.data);
    }

    public render() {
      return <App {...this.props} apolloClient={this.apolloClient} />;
    }
  };
};
