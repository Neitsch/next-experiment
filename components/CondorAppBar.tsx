import AppBar from "@material-ui/core/AppBar";
import Button from "@material-ui/core/Button";
import IconButton from "@material-ui/core/IconButton";
import Menu from "@material-ui/core/Menu";
import MenuItem from "@material-ui/core/MenuItem";
import { withStyles, WithStyles } from "@material-ui/core/styles";
import Toolbar from "@material-ui/core/Toolbar";
import Typography from "@material-ui/core/Typography";
import AccountCircle from "@material-ui/icons/AccountCircle";
import MenuIcon from "@material-ui/icons/Menu";
import gql from "graphql-tag";
import React from "react";
import { Query } from "react-apollo";

import { Router } from "../lib/routes";

const styles = {
  flex: {
    flex: 1,
  },
};

interface IProps {
  isAuthenticated: boolean;
}

type Props = IProps & WithStyles<typeof styles>;

interface IState {
  openAccountMenu?: HTMLElement;
}

const APP_BAR_DATA = gql`
  query AppBarData {
    user {
      username
    }
  }
`;

class CondorAppBar extends React.Component<Props, IState> {
  constructor(props: Props) {
    super(props);
    this.state = {
      openAccountMenu: null,
    };
  }
  public render() {
    const { classes, isAuthenticated } = this.props;
    return (
      <AppBar position="static">
        <Toolbar>
          <IconButton aria-label="Menu">
            <MenuIcon />
          </IconButton>
          <Typography variant="title" color="inherit" className={classes.flex}>
            Condor Club
          </Typography>
          {isAuthenticated ? (
            <div>
              <IconButton
                aria-owns={"menu-appbar"}
                aria-haspopup="true"
                onClick={this.handleAccountMenu}
                color="inherit"
                id="accountButton"
              >
                <AccountCircle />
              </IconButton>
              {this.state.openAccountMenu ? (
                <Query query={APP_BAR_DATA} id="query">
                  {({ data }) => {
                    return data && data.user ? (
                      <Menu
                        id="menu-appbar"
                        anchorEl={this.state.openAccountMenu}
                        anchorOrigin={{
                          horizontal: "right",
                          vertical: "top",
                        }}
                        transformOrigin={{
                          horizontal: "right",
                          vertical: "top",
                        }}
                        open={true}
                        onClose={() =>
                          this.setState(() => ({ openAccountMenu: null }))
                        }
                      >
                        <MenuItem>
                          Hey,{" "}
                          {data.user.username
                            ? data.user.username
                            : "Anonymous"}
                        </MenuItem>
                        <MenuItem id="logout" onClick={this.logout}>
                          Sign Out
                        </MenuItem>
                      </Menu>
                    ) : null;
                  }}
                </Query>
              ) : null}
            </div>
          ) : (
            <Button id="login" color="inherit">
              Login
            </Button>
          )}
        </Toolbar>
      </AppBar>
    );
  }
  private handleAccountMenu = (event: React.SyntheticEvent<HTMLElement>) => {
    const target = event.currentTarget;
    this.setState(() => ({
      openAccountMenu: target,
    }));
  };
  private logout = () => {
    Router.pushRoute("/auth/sign-out");
  };
}

export default withStyles(styles)(CondorAppBar);
