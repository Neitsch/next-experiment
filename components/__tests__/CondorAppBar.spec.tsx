import { createShallow } from "@material-ui/core/test-utils";
import React from "react";

import { Router } from "../../lib/routes";
import CondorAppBar from "../CondorAppBar";

jest.unmock("../CondorAppBar");
jest.mock("graphql-tag", () =>
  jest.fn().mockImplementation(a => ({
    data: a,
  })),
);

describe("AppBar", () => {
  let shallow;
  beforeEach(() => {
    shallow = createShallow({ dive: true });
  });
  describe("unauthenticated", () => {
    let AppBarRendered = null;
    beforeEach(() => {
      const AppBarComponent = <CondorAppBar isAuthenticated={false} />;
      AppBarRendered = shallow(AppBarComponent);
    });
    it("renders", () => {
      expect(AppBarRendered).toMatchSnapshot();
    });
  });
  describe("authenticated", () => {
    let AppBarRendered = null;
    beforeEach(() => {
      const AppBarComponent = <CondorAppBar isAuthenticated={true} />;
      AppBarRendered = shallow(AppBarComponent);
    });
    it("renders", () => {
      expect(AppBarRendered).toMatchSnapshot();
    });
    it("Open Menu", () => {
      const method = AppBarRendered.find("#accountButton").props().onClick;
      const anchorElem = jest.fn();
      method({ currentTarget: anchorElem });
      expect(AppBarRendered.state("openAccountMenu")).toEqual(anchorElem);
    });
    it("Logout", () => {
      AppBarRendered.setState({
        openAccountMenu: AppBarRendered, // Dummy element
      });
      const queryComponent = AppBarRendered.find("#query")
        .props()
        .children({ data: { user: { username: "Test Username" } } });
      const queryRender = shallow(queryComponent);
      expect(queryRender).toMatchSnapshot();

      const queryComponentNoUsername = AppBarRendered.find("#query")
        .props()
        .children({ data: { user: {} } });
      const queryRenderNoUsername = shallow(queryComponentNoUsername);
      expect(queryRenderNoUsername).toMatchSnapshot();

      expect(
        AppBarRendered.find("#query")
          .props()
          .children({ data: {} }),
      ).toBeNull();

      const method = queryRender.find("#logout").props().onClick;
      method();
      expect(Router.pushRoute).toHaveBeenCalledTimes(1);

      const settings = queryRender.find("#settingsButton").props().onClick;
      settings();
      expect(Router.pushRoute).toHaveBeenCalledTimes(2);

      const onCloseMethod = queryRender.find("#menu-appbar").props().onClose;
      expect(AppBarRendered.state("openAccountMenu")).not.toBeNull();
      onCloseMethod();
      expect(AppBarRendered.state("openAccountMenu")).toBeNull();
      expect(AppBarRendered.exists("#query")).toBeFalsy();
    });
  });
});
