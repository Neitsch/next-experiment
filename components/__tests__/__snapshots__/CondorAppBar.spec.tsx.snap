// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppBar authenticated Logout 1`] = `
<WithStyles(Popover)
  PaperProps={
    Object {
      "classes": Object {
        "root": "MuiMenu-paper-2",
      },
    }
  }
  anchorEl={
    <WithStyles(AppBar)
      position="static"
    >
      <WithStyles(Toolbar)>
        <WithStyles(IconButton)
          aria-label="Menu"
        >
          <pure(Menu) />
        </WithStyles(IconButton)>
        <WithStyles(Typography)
          className="CondorAppBar-flex-1"
          color="inherit"
          variant="title"
        >
          Condor Club
        </WithStyles(Typography)>
        <div>
          <WithStyles(IconButton)
            aria-haspopup="true"
            aria-owns="menu-appbar"
            color="inherit"
            id="accountButton"
            onClick={[Function]}
          >
            <pure(AccountCircle) />
          </WithStyles(IconButton)>
          <mockConstructor
            id="query"
            query={
              Object {
                "data": Array [
                  "
  query AppBarData {
    user {
      username
    }
  }
",
                ],
              }
            }
          >
            <Component />
          </mockConstructor>
        </div>
      </WithStyles(Toolbar)>
    </WithStyles(AppBar)>
  }
  anchorOrigin={
    Object {
      "horizontal": "right",
      "vertical": "top",
    }
  }
  getContentAnchorEl={[Function]}
  id="menu-appbar"
  onClose={[Function]}
  onEnter={[Function]}
  open={true}
  transformOrigin={
    Object {
      "horizontal": "right",
      "vertical": "top",
    }
  }
  transitionDuration="auto"
>
  <MenuList
    onKeyDown={[Function]}
  >
    <WithStyles(MenuItem)>
      Hey,
       
      Test Username
    </WithStyles(MenuItem)>
    <WithStyles(Divider) />
    <WithStyles(MenuItem)
      id="settingsButton"
      onClick={[Function]}
    >
      Settings
    </WithStyles(MenuItem)>
    <WithStyles(MenuItem)
      id="logout"
      onClick={[Function]}
    >
      Sign Out
    </WithStyles(MenuItem)>
  </MenuList>
</WithStyles(Popover)>
`;

exports[`AppBar authenticated Logout 2`] = `
<WithStyles(Popover)
  PaperProps={
    Object {
      "classes": Object {
        "root": "MuiMenu-paper-2",
      },
    }
  }
  anchorEl={
    <WithStyles(AppBar)
      position="static"
    >
      <WithStyles(Toolbar)>
        <WithStyles(IconButton)
          aria-label="Menu"
        >
          <pure(Menu) />
        </WithStyles(IconButton)>
        <WithStyles(Typography)
          className="CondorAppBar-flex-1"
          color="inherit"
          variant="title"
        >
          Condor Club
        </WithStyles(Typography)>
        <div>
          <WithStyles(IconButton)
            aria-haspopup="true"
            aria-owns="menu-appbar"
            color="inherit"
            id="accountButton"
            onClick={[Function]}
          >
            <pure(AccountCircle) />
          </WithStyles(IconButton)>
          <mockConstructor
            id="query"
            query={
              Object {
                "data": Array [
                  "
  query AppBarData {
    user {
      username
    }
  }
",
                ],
              }
            }
          >
            <Component />
          </mockConstructor>
        </div>
      </WithStyles(Toolbar)>
    </WithStyles(AppBar)>
  }
  anchorOrigin={
    Object {
      "horizontal": "right",
      "vertical": "top",
    }
  }
  getContentAnchorEl={[Function]}
  id="menu-appbar"
  onClose={[Function]}
  onEnter={[Function]}
  open={true}
  transformOrigin={
    Object {
      "horizontal": "right",
      "vertical": "top",
    }
  }
  transitionDuration="auto"
>
  <MenuList
    onKeyDown={[Function]}
  >
    <WithStyles(MenuItem)>
      Hey,
       
      Anonymous
    </WithStyles(MenuItem)>
    <WithStyles(Divider) />
    <WithStyles(MenuItem)
      id="settingsButton"
      onClick={[Function]}
    >
      Settings
    </WithStyles(MenuItem)>
    <WithStyles(MenuItem)
      id="logout"
      onClick={[Function]}
    >
      Sign Out
    </WithStyles(MenuItem)>
  </MenuList>
</WithStyles(Popover)>
`;

exports[`AppBar authenticated renders 1`] = `
<WithStyles(AppBar)
  position="static"
>
  <WithStyles(Toolbar)>
    <WithStyles(IconButton)
      aria-label="Menu"
    >
      <pure(Menu) />
    </WithStyles(IconButton)>
    <WithStyles(Typography)
      className="CondorAppBar-flex-1"
      color="inherit"
      variant="title"
    >
      Condor Club
    </WithStyles(Typography)>
    <div>
      <WithStyles(IconButton)
        aria-haspopup="true"
        aria-owns="menu-appbar"
        color="inherit"
        id="accountButton"
        onClick={[Function]}
      >
        <pure(AccountCircle) />
      </WithStyles(IconButton)>
    </div>
  </WithStyles(Toolbar)>
</WithStyles(AppBar)>
`;

exports[`AppBar unauthenticated renders 1`] = `
<WithStyles(AppBar)
  position="static"
>
  <WithStyles(Toolbar)>
    <WithStyles(IconButton)
      aria-label="Menu"
    >
      <pure(Menu) />
    </WithStyles(IconButton)>
    <WithStyles(Typography)
      className="CondorAppBar-flex-1"
      color="inherit"
      variant="title"
    >
      Condor Club
    </WithStyles(Typography)>
    <WithStyles(Button)
      color="inherit"
      id="login"
    >
      Login
    </WithStyles(Button)>
  </WithStyles(Toolbar)>
</WithStyles(AppBar)>
`;
